[Provider] Upload the list of employees

HlibBondarev opened the pull request:
**I. Algorithm**
1) Check whether there is an Individual with the passed Rnokpp in DB.
2) If not, we have to create a new Individual and attach it to the current Provider with passed Id (through Official and Position entities). 
3) If Individual exists, we have to check whether Individual attached to the current Provider.
(We check presence Individual in Official for the current Provider and also DismissalOrder property is null)
4) If not, we have to attach Individual to the current Provider (through Official and Position entities).

**II. Realization**
1) Added UploadEmployeesForProvider() method to ProviderService class.
This method includes 3 auxiliary private methods - 
CheckListOfEmployeesForUploading (for checking the passed employee list), AddIndividualsToDb (for adding Individuals from the employee list to DB) and  FillDbWithNewEmployeesOnPositions (this method contains the main logic of adding new employees to DB for  Provider with the certain Id).
2) Added Upload action to ProviderController class.
3) Added unit tests for checking of UploadEmployeesForProvider() method of ProviderService and Upload() action of ProviderController.

**III. Code updates**
1) Added UploadEmployeeRequestDto class for uploading new employees.
2) Added UploadEmployeeResponse class for response.
3) Changed IProviderService interface - added method signature: 
Task UploadEmployeesForProvider(Guid id, UploadEmployeeRequestDto[] data);
4) Changed ProviderService class:
- injected repositories for Individual, Official and Position to the service;
- added async method: Task UploadEmployeesForProvider(Guid id, UploadEmployeeRequestDto[] data).
5) Changed ProviderServiceV2 class: injected repositories for Individual, Official and Position to the service.
6) Changed MappingProfile class: added mapping UploadEmployeeRequestDto => Individual and Individual => UploadEmployeeRequestDto.
7) Changed Constants class: added constants - MaxIndividualNameLength, MinIndividualNameLength and MaxNumberOfEmployeesToUpload.
8) Changed Position class: added default value (string.Empty) for properties - GenitiveName and ClassifierType. 
9) Changed User class: reodered properties and added comment - TODO. 
10) Changed ProviderController class: added action - 
async Task<IActionResult> Upload(Guid id, [FromBody] UploadEmployeeRequestDto [] uploadEployees). 
11) Changed tests in classes ProviderServiceTests, ProviderServiceV2, ProviderServiceTests and  ProviderServiceUpdate - added mocks of repositories for Individual, Official and Position.
12) Created UploadEmployeeDtoGenerator, OfficeGenerator and PositionGenerator for generating instances of the corresponding entities.
13) Added tests for UploadEmployeesForProvider() method of ProviderService to ProviderServiceTests class.
14) Added test for Upload() action of ProviderController in ProviderControllerTests class.

HlibBondarev committed:
1) Added UploadEmployeeDto class for uploading new employees.
2) Changed IProviderService interface - added method signature:
Task UploadEmployeesForProvider(Guid id, UploadEmployeeDto[] data);
3) Changed ProviderService class:
- injected repositories for Individual, Official and Position to the service;
- added async method: Task UploadEmployeesForProvider(Guid id, UploadEmployeeDto[] data).
4) Changed ProviderServiceV2 class: injected repositories for Individual, Official and Position to the service.
5) Changed MappingProfile class: added mapping UploadEmployeeDto => Individual and Individual => UploadEmployeeDto.
6) Changed Constants class: added constants - MaxIndividualNameLength, MinIndividualNameLength and MaxNumberOfEmployeesToUpload.
7) Changed Position class: added default value (string.Empty) for properties - GenitiveName and ClassifierType.
8) Changed User class: reodered properties and added comment - TODO.
9) Added ProviderController class: added action - async Task<IActionResult> Upload(Guid id, [FromBody] UploadEmployeeDto[] uploadEployees).
10) Changed tests in classes ProviderServiceTests, ProviderServiceV2, ProviderServiceTests and  ProviderServiceUpdate -
added mocks of repositories for Individual, Official and Position.

HlibBondarev committed:
Fixed SonarCloud issues

HlibBondarev committed:
Refactored UploadEmployeesForProvider() method of ProviderService class: extracted CheckListOfEmployeesForUploading() method for checking employees.

HlibBondarev committed:
1) Renamed UploadEmployeeDto class to UploadEmployeeRequestDto.
2) Created UploadEmployeeResponseDto class.
3) Changed signature of method Task<UploadEmployeeResponseDto> UploadEmployeesForProvider(Guid id, UploadEmployeeRequestDto[] data) in IProviderService interface and ProviderService class.
4) Changed Upload() action of ProviderController class.
5) Added tests for UploadEmployeesForProvider() method of ProviderService class to ProviderServiceTests class.
6) Created UploadEmployeeDtoGenerator class for generating UploadEmployeeDto instances.

HlibBondarev committed:
Changed ProviderController class.

HlibBondarev committed:
Refactored UploadEmployeesForProvider() method of ProviderService class.

HlibBondarev committed:
Fixed a bug in UploadEmployeesForProvider() method of ProviderService class.

HlibBondarev committed:
1) Renamed UploadEmployeeResponseDto class to UploadEmployeeResponse.
2) Refactored UploadEmployeesForProvider() method of ProviderService class:
extracted CheckListOfEmployeesForUploading() methods - AddIndividualsToDb() and FillDbWithNewEmployeesOnPositions().

HlibBondarev committed:
1) Added data generators - OfficialGenerator and PositionGenerator classes.
2) Added tests for UploadEmployeesForProvider() method of ProviderService class to ProviderServiceTests class.

HlibBondarev committed:
1) Changed tests for UploadEmployeesForProvider() method of ProviderService class - added:
individualRepositoryMock.VerifyAll();
officialRepositoryMock.VerifyAll();
positionRepositoryMock.VerifyAll();
2) Added tests for Upload() Action of ProviderController to ProviderControllerTests class.

HlibBondarev committed:
Added test for Upload() Action of ProviderController in ProviderControllerTests class -
public async Task Upload_WhenServisThrowsException_ThrowsInvalidOperationException(int entitiesCount).

h4wk13 commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/ProviderServices/ProviderService.cs:
```diff
@@ -1013,4 +1049,115 @@ private async Task<bool> ExistsAnotherProviderWithTheSameEdrpouIpn(ProviderUpdat
+    {
+        _ = data ?? throw new ArgumentNullException(nameof(data));
+
+        logger.LogInformation("Upload employees for provider was started.");
```
It does not look like info, it looks like `Debug`

HlibBondarev replied:
Ok, replaced logger.LogInformation() with logger.LogDebug()

h4wk13 commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/ProviderServices/ProviderService.cs:
```diff
@@ -1013,4 +1049,115 @@ private async Task<bool> ExistsAnotherProviderWithTheSameEdrpouIpn(ProviderUpdat
+        {
+            var errorMessage = $"The number of entries should not exceed {Constants.MaxNumberOfEmployeesToUpload}.";
+            logger.LogError("The number of entries should not exceed {MaxNumberOfEmployeesToUpload}.", Constants.MaxNumberOfEmployeesToUpload);
+            throw new InvalidOperationException(errorMessage);
```
`ArgumentOutOfRangeException` ?

HlibBondarev replied:
Ok, replaced InvalidOperationException with ArgumentOutOfRangeException

h4wk13 commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/ProviderServices/ProviderService.cs:
```diff
@@ -1013,4 +1049,115 @@ private async Task<bool> ExistsAnotherProviderWithTheSameEdrpouIpn(ProviderUpdat
+        {
+            var errorMessage = "The number of entries to upload should be greater than 0.";
+            logger.LogError(errorMessage);
+            throw new InvalidOperationException(errorMessage);
```
`ArgumentOutOfRangeException` ?

HlibBondarev replied:
Ok, replaced InvalidOperationException with ArgumentOutOfRangeException

h4wk13 commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/ProviderServices/ProviderService.cs:
```diff
@@ -1013,4 +1049,115 @@ private async Task<bool> ExistsAnotherProviderWithTheSameEdrpouIpn(ProviderUpdat
+    {
+        // Dictionary for uploading employees
+        var uploadDictionary = new Dictionary<Guid, UploadEmployeeRequestDto>();
+        var existingIndividuals = (await individualRepository.GetByFilter(i => data.Select(e => e.Rnokpp).Contains(i.Rnokpp))
```
please move `data.Select(e => e.Rnokpp)` into a local variable 
And how long that list might be ?

HlibBondarev replied:
Ok, moved data.Select(e => e.Rnokpp) into a local variable.
The number of members in this list is limited by a constant - MaxNumberOfEmployeesToUpload.
According to requirments MaxNumberOfEmployeesToUpload = 100

h4wk13 commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/ProviderServices/ProviderService.cs:
```diff
@@ -1013,4 +1049,115 @@ private async Task<bool> ExistsAnotherProviderWithTheSameEdrpouIpn(ProviderUpdat
+                                                                          .Select(i => new { i.Rnokpp, i.Id })
+                                                                          .ToDictionary(e => e.Rnokpp);
+        
+        // Circle to add individuals to DB and populate the Dictionary for uploading employees
```
`Circle` -> `loop` ?

HlibBondarev replied:
Ok, Changed Circle -> loop

h4wk13 reviewed (changes requested):
As for me - it is not clear what is going on in many places in `ProviderService`.
I would ask for refactoring of it.

HlibBondarev committed:
Edited code according to requested changes in PR:
1) Changed CheckListOfEmployeesForUploading() method of  ProviderService class - used ArgumentOutOfRangeException intead of InvalidOperationException.
2) Changed Upload() action of ProviderController - added catch block (catch (ArgumentOutOfRangeException ex)).
3) Added tests to ProviderServiceTests and ProviderControllerTests classes.

HlibBondarev committed:
Changed private method - FillDbWithNewEmployeesOnPositions():
1) Replaced type of existingOfficialsForProvider var -
from ILookup<Guid, Position> with ILookup<Guid, String>.
2) Simplifyed condition in the loop for filling the DB with new employees on certain positions.

HlibBondarev committed:
Replaced logger.LogInformation() with logger.LogDebug() in UploadEmployeesForProvider() method of ProviderService class.

HlibBondarev committed:
Merge branch 'develop' into HlibBondarev/provider_upload_list_of_employees

HlibBondarev committed:
Fixed conflicts with develop branch
